### PR TITLE
Added Jacobians for Rot3::xyz and related conversions to euler angles

### PR DESCRIPTION
**Pull request by Eskilade**
Hi,

This PR adds the remaining Jacobians for conversions between Rot3 and euler angles ([first PR here](https://github.com/borglab/gtsam/pull/499) ). Some of the expressions below are a bit hairy, so please feel free to suggest a different way of doing it :)

### CODE REVIEW

**Comment by jlblancoc** [comment]
Aeesome!

**Comment by jlblancoc** [comment]
Perhaps it would be convenient to add unit tests covering ypr near the extreme +-pi and +-pi/2 values?

File: gtsam/geometry/Rot3.cpp [hunk 1]
```diff
@@ -158,21 +158,70 @@ Point3 Rot3::column(int index) const{
...
+    boost::tie(I, q) = RQ(matrix());
   return q;
 }
 
 /* ************************************************************************* */
-Vector3 Rot3::ypr() const {
-  Vector3 q = xyz();
+Vector3 Rot3::ypr(OptionalJacobian<3, 3> H) const {
+  Vector3 q = xyz(H);
+  if (H) H->row(0).swap(H->row(2));
```
Comment (dellaert) [inline gtsam/geometry/Rot3.cpp:183]: nice !

File: gtsam/geometry/Rot3.cpp [hunk 2]
```diff
@@ -158,21 +158,70 @@ Point3 Rot3::column(int index) const{
...
+    const auto m = matrix();
+#ifdef GTSAM_USE_QUATERNIONS
+    SO3{m}.vec(mH);
+#else
+    rot_.vec(mH);
+#endif
+
+    Matrix39 qHm;
+    boost::tie(I, q) = RQ(m, qHm);
+    *H = qHm * mH;
```
Comment (dellaert) [inline gtsam/geometry/Rot3.cpp:174]: So, this is of course perfectly fine. However, if you care about performance, this is (a) 81 mults, and (b) we could use of the fact that mH is super-sparse. In particular, SO3.cpp says
```
    mH << R * P3.block<3, 3>(0, 0), R * P3.block<3, 3>(3, 0),
        R * P3.block<3, 3>(6, 0);
```
Maybe add as a TODO. In fact I suspect the 3*3 matrix of derivatives might yield not *too* complicated symbolic expressions, and we do a lot of compute here (including multiplying a lot of things with zero). Might be worth some mathematica exploration.

**Review by dellaert — APPROVED** [review]
Great! I'm OK to approve as is, but I think the computation is quite heavy and could probably be improved quite a bit. Still, we did not have this before, and it might not make a huge difference in an application.

**Comment by Eskilade** [comment]
> Perhaps it would be convenient to add unit tests covering ypr near the extreme +-pi and +-pi/2 values?
Ok, I'll add more tests. However, the derivative breaks down for pitch values +- pi / 2. How do you prefer to handle that? Do you use nans?

**Comment by jlblancoc** [comment]
> However, the derivative breaks down for pitch values +- pi / 2. How do you prefer to handle that? Do you use nans?

That was my point, bringing out that discussion :-) 
Let's see what @dellaert thinks, but I feel leaned to say that the unit test must pass for values near the edge, and raise an exception if exactly (within an epsilon) of the edges if the derivative (not even the numerical one) is really not well-defined there.

**Comment by dellaert** [comment]
> > However, the derivative breaks down for pitch values +- pi / 2. How do you prefer to handle that? Do you use nans?
> 
> That was my point, bringing out that discussion :-)
> Let's see what @dellaert thinks, but I feel leaned to say that the unit test must pass for values near the edge, and raise an exception if exactly (within an epsilon) of the edges if the derivative (not even the numerical one) is really not well-defined there.

I think an exception is a good idea. There is a singularity, for sure.

**Comment by Eskilade** [comment]
Hmm, I'm a bit unsure about which epsilon value we should use... If I set roll and yaw to zero, and if 
a) is the largest entry in the difference between the numerical derivative and the calculated derivative in absolute value 
b) the entry in the calculated derivative corresponding to (a)

then I get :
 * a) ~3e-08  b) ~10     for pitch = pi/2 - 1e-1
 * a) ~3e-05  b) ~100   for pitch = pi/2 - 1e-2
 * a) ~3e-02  b) ~1e3   for pitch = pi/2 - 1e-3
 * a) ~30       b) ~1e4   for pitch = pi/2 - 1e-4
 * a) ~2e4     b) ~1e5   for pitch = pi/2 - 1e-5

What do you think? I hope the numbers are understandable

**Comment by jlblancoc** [comment]
> * a) ~3e-05  b) ~100   for pitch = pi/2 - 0.01

Thanks for the great summary and the time collecting the results!
I think "pi/2 - 0.01" would be safe enough. Don't set the unit test threshold too tight (e.g. if your error is ~3e-05 , mark the limit like 1 order of magnitude larger than that) since different CPU architectures may give slightly different numerical results for the same code and make the unit test to fail. 
Thanks!

**Comment by dellaert** [comment]
@Eskilade let me know when ready to merge....

**Comment by Eskilade** [comment]
Yupp, ok for my part ! Btw, does this merit a mention in contributors or something similar ? :)

**Comment by dellaert** [comment]
Of course, feel free to add your name, and then I'll merge :-)

**Comment by Eskilade** [comment]
Where do I add it ? Like in which file :)

**Comment by varunagrawal** [comment]
You can add your name here: https://github.com/borglab/gtsam.org/blob/master/about.md

**Comment by Eskilade** [comment]
Alright, but since that's a different repo, then you can go ahead and merge this PR :)

**Comment by lanyouzibetty** [comment]
@Eskilade Thanks for your awesome work. Can you tell me how to derive the jacobians for conversions between Rot3 and euler angles? I have not deduced it, thanks!

**Comment by Eskilade** [comment]
@lanyouzibetty, have a look here : https://github.com/Eskilade/orient/blob/master/documentation/conversion_formulas.pdf :)

**Comment by lanyouzibetty** [comment]
@Eskilade Thanks very much !

**Pull request merged**